pyccel/parser/syntactic.py

```diff
@@ -1107,6 +1122,10 @@ def _visit_For(self, stmt):
         iterable = self._visit(stmt.iter)
         body = self._visit(stmt.body)
 
+        if any(isinstance(l, (FunctionDef, ClassDef, Import)) for l in body):
+            errors.report("Unexpected object in a while block (function, class or import)",
+                    symbol=stmt, severity='error')
+
```

Comment:
It makes more sense to have the checks in the semantics stage, where we raise this error in
`_visit_FunctionDef`, `_visit_ClassDef`, and `_visit_Import` if the methods are called, and it is more efficient,
because here we go over the body every time to check for a special case, contrary to the other method which will not do the check unless those methods are called.

Reply:
But if we want to have these in the semantic stage then we either have to add a `functions` argument to `With`, `While` and `For` or we need to add an inheritance check in `_visit_FunctionDef`, `_visit_ClassDef`, and `_visit_Import`. I suspect such a check might be relatively complex.

The simplest solution I can think of is to check the `direct_user_nodes` and exclude acceptable enclosing nodes (`Module`/`FunctionDef`/`FunctionCall`). Given the speed of `any` I am not convinced that this would lead to faster runtimes. Do you see a simpler check?

Reply:
Fixed : c40da083d

Reply:
I'm not a huge fan of the idea of creating a `handle_functiondef` which completely replaces `_visit_FunctionDef` as it means we no longer use the function `_visit` which does a few extra commonly forgotten steps like updating `self.current_ast_node` and setting the ast of the resulting object

Reply:
what does `get_direct_user_nodes` do? does it give the parent nodes or what exactly?

Reply:
Improved docs: eb76720f1

Reply:
Okay, it seems that the cost of this function is O(1) as most objects have one user node, so it should be fine

Reply:
Most objects except the only ones we are applying it to. A `FunctionDef` usually has lots of user nodes as it is used whenever it is called

Reply:
But I don't think that this loop should be worse than the one in `For`/`While`/`With`

Reply:
Okay, it should be fine